tools/aux/tpm2_print.c

```diff
@@ -0,0 +1,336 @@
+    tpm2_util_hexdump((const UINT8*)&magic, sizeof(UINT32), true);
+    tpm2_tool_output("\n");
+    magic = tpm2_util_ntoh_32(magic); // finally, convert endianness
+
+    // check magic
+    if (magic != TPM2_GENERATED_VALUE) {
+        LOG_ERR("%s: Bad magic", file_path);
+        goto error;
+    }
+
+    UINT16 type;
+    res = files_read_bytes(fd, (UINT8*)&type, sizeof(UINT16));
+    if (!res) {
+        goto read_error;
+    }
+    print_yaml_indent(indent_count);
+    tpm2_tool_output("type: ");
+    tpm2_util_hexdump((const UINT8*)&type, sizeof(UINT16), true);
+    tpm2_tool_output("\n");
+    type = tpm2_util_ntoh_16(type); // finally, convert endianness
+
+    print_yaml_indent(indent_count);
+    tpm2_tool_output("qualifiedSigner: ");
+    res = tpm2_util_print_tpm2b_file(fd);
+    tpm2_tool_output("\n");
+    if (!res) {
+        goto read_error;
+    }
+
+    print_yaml_indent(indent_count);
+    tpm2_tool_output("extraData: ");
+    res = tpm2_util_print_tpm2b_file(fd);
+    tpm2_tool_output("\n");
+    if (!res) {
+        goto read_error;
+    }
+
+    print_yaml_indent(indent_count);
+    tpm2_tool_output("clockInfo:\n");
+    res = print_clock_info(fd, indent_count + 1);
+    if (!res) {
+        goto error;
+    }
+
+    print_yaml_indent(indent_count);
+    tpm2_tool_output("firmwareVersion: ");
+    res = tpm2_util_hexdump_file(fd, sizeof(UINT64), true);
+    tpm2_tool_output("\n");
+    if (!res) {
+        goto read_error;
+    }
+
+    switch(type) {
+    case TPM2_ST_ATTEST_QUOTE:
+        print_yaml_indent(indent_count);
+        tpm2_tool_output("attested:\n");
+        print_yaml_indent(indent_count + 1);
+        tpm2_tool_output("quote:\n");
+        res = print_TPMS_QUOTE_INFO(fd, indent_count + 2);
+        if (!res) {
+            goto error;
+        }
+        break;
+
+    default:
+        LOG_ERR("%s: Cannot print unsupported type 0x%x", file_path, (unsigned int)type);
+        goto error;
+    }
+
+    // success
+    return true;
+
+read_error:
+    FILE_2_SHORT_ERROR();
+error:
+    return false;
+}
+
+static bool on_option(char key, char *value) {
+    switch (key) {
+    case 't':
+        if (strcmp(value, "TPMS_ATTEST") == 0) {
+            file_type = file_type_TPMS_ATTEST;
+        }
+        else {
+            LOG_ERR("Invalid type specified. Only TPMS_ATTEST is presently supported.\n");
+            return false;
+        }
+        break;
+    case 'f':
+        file_path = value;
+        break;
+    }
+
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+    static const struct option topts[] = {
+        { "type",        required_argument, NULL, 't' },
+        { "file",        optional_argument, NULL, 'f' },
+    };
+
+    *opts = tpm2_options_new("t:f:", ARRAY_LEN(topts), topts,
+        on_option, NULL);
+
+    return *opts != NULL;
+}
+
+int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
+    UNUSED(sapi_context);
```

Comment:
Not related with your patch, but this tool could be used even if there's no TPM available. But currently it will fail even when the SAPI context is not used. I've filled issue #737 for that.

Reply:
Yeah perhaps we need a way in onstart to indicate that they don't need a SAPI. tpm2_rc_decode has the same issue, and could go to aux as well at some point.

Reply:
I noticed that as well, but it hasn't been an issue for us yet. I'd prefer to compile out the SAPI initialization instead of a runtime flag, so that we can remove the library dependency as well.

Reply:
@harisokanovic you'd still need the headers for data types, so it would only remove the link dependency.